src/main/java/org/jenkinsci/plugins/workflow/log/DelayBufferedOutputStream.java

```diff
@@ -62,85 +105,97 @@ private Tuning() {}
-    private void flushBuffer() throws IOException {
-        ThreadLocal<Boolean> enableFlush = ((FlushControlledOutputStream) out).enableFlush;
-        boolean orig = enableFlush.get();
-        enableFlush.set(false);
-        try {
-            flush();
-        } finally {
-            enableFlush.set(orig);
-        }
+    @Override public void write(byte[] b, int off, int len) throws IOException {
+        buf.write(b, off, len);
+    }
+
+    @Override public void flush() throws IOException {
+        buf.drain();
+        super.flush();
+    }
+
+    private void reschedule() {
+        Timer.get().schedule(new Drain(this), recurrencePeriod, TimeUnit.MILLISECONDS);
+        recurrencePeriod = Math.min((long) (recurrencePeriod * tuning.recurrencePeriodBackoff), tuning.maxRecurrencePeriod);
     }
 
-    void flushAndReschedule() {
+    void drainAndReschedule() {
         // TODO as an optimization, avoid flushing the buffer if it was recently flushed anyway due to filling up
         try {
-            flushBuffer();
+            buf.drain();
         } catch (IOException x) {
             LOGGER.log(Level.FINE, null, x);
         }
         reschedule();
     }
 
-    @SuppressWarnings("FinalizeDeclaration") // not ideal, but PhantomReference is more of a hassle
-    @Override protected void finalize() throws Throwable {
-        super.finalize();
-        // Odd that this is not the default behavior for BufferedOutputStream.
-        flush();
-    }
-
-    private static final class Flush implements Runnable {
+    private static final class Drain implements Runnable {
 
         /** Since there is no explicit close event, just keep flushing periodically until the stream is collected. */
         private final Reference<DelayBufferedOutputStream> osr;
 
-        Flush(DelayBufferedOutputStream os) {
+        Drain(DelayBufferedOutputStream os) {
             osr = new WeakReference<>(os);
         }
 
         @Override public void run() {
             DelayBufferedOutputStream os = osr.get();
             if (os != null) {
-                os.flushAndReschedule();
+                os.drainAndReschedule();
             }
         }
 
     }
 
-    /** @see DelayBufferedOutputStream#flushBuffer */
-    private static final class FlushControlledOutputStream extends FilterOutputStream {
-
-        private final ThreadLocal<Boolean> enableFlush = new ThreadLocal<Boolean>() {
-            @Override protected Boolean initialValue() {
-                return true;
-            }
-        };
-
-        FlushControlledOutputStream(OutputStream out) {
-            super(out);
+    /**
+     * Flushes streams prior to garbage collection.
+     * In Java 9+ could use {@code java.util.Cleaner} instead.
+     */
+    private static final class FlushRef extends PhantomReference<DelayBufferedOutputStream> {
```

Comment:
Seriously hacky, but I *think* this will work IIUC what it's doing.

Makes me wonder why Remoting doesn't have a way to register a callback to execute before disposing the lease on an object.